frontend/catalyst/compiler.py

```diff
@@ -268,6 +269,63 @@ def get_output_filename(infile):
         return infile.replace(".ll", ".o")
 
 
+class WrapperToCatchExceptions:
+    """Create a wrapper around the generated function to catch runtime exceptions."""
+
+    def __init__(self, params, name):
+        self.params = params
+        self.name = name
+
+    _include = sysconfig.get_paths()["include"]
+    _executable = "clang++"
+    _default_flags = ["-fPIC", "-x", "c++", f"-I{_include}"]
+
+    @staticmethod
+    def get_output_filename(infile):
+        """Get a name for the wrapper object file."""
+        # Technically we are just using this to get the whole path
+        if not infile.endswith(".mlir"):
+            raise ValueError("Input is not an MLIR file.")
+        return infile.replace(".mlir", ".wrap.o")
+
+    def run(self, infile, outfile=None, options=None):
+        """Compile code to generate a wrapper around the user's program."""
+        name = self.name
+        params = [f"void* arg{x}" for x in range(self.params)]
+        args = [f"arg{x}" for x in range(self.params)]
+        params_string = ", ".join(params)
+        args_string = ", ".join(args)
+        template = f"""
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <stdexcept>
+#include <stdarg.h>
+
+extern "C" {{
+  extern void _mlir_ciface_{name} ({ params_string });
+  void wrapper_mlir_ciface_{name} ({ params_string });
+}}
+
+void wrapper_mlir_ciface_{name} ({ params_string }) {{
+  try {{
+    _mlir_ciface_{name} ({ args_string });
+  }} catch (const std::exception &ex) {{
+    PyErr_SetString(PyExc_RuntimeError, ex.what());
+  }} catch (const std::logic_error &le) {{
+    PyErr_SetString(PyExc_RuntimeError, le.what());
+  }} catch (const std::range_error &re) {{
+    PyErr_SetString(PyExc_RuntimeError, re.what());
```

Comment:
It is recommended to order the catch blocks from most specific to least specific,
```suggestion
  }} catch (const std::logic_error &le) {{
    PyErr_SetString(PyExc_RuntimeError, le.what());
  }} catch (const std::range_error &re) {{
    PyErr_SetString(PyExc_RuntimeError, re.what());
  }} catch (const std::exception &ex) {{
    PyErr_SetString(PyExc_RuntimeError, ex.what());
```